### ISSUE TITLE
Improve the ergonomy of dealing with `.eh_frame` and `.debug_frame`

### ISSUE THREAD

**Issue by koute** [gimli-rs/gimli#296]
Often we might not care whenever we're dealing with `.eh_frame` or `.debug_frame` and we just want to treat those two the same.

For dealing with `.eh_frame` we have to use the `EhFrame` type while for `.debug_frame` we use `DebugFrame`. Those two are then used as type parameters in `UninitializedUnwindContext`, `InitializedUnwindContext`, `CommonInformationEntry`, etc.

This is a problem when we just don't care which one it is, and we just want to dynamically dispatch to the relevant one at runtime based on whenever an `.eh_frame` or a `.debug_frame` was loaded.

Another problem is that manually caching the unwind contexts can be a little bit awkward, especially since the types are different.

Such an API might perhaps be a little too high level and specific to my use case, but something like this could be used as an inspiration to help think about what could be done here: (Copied from [here](https://github.com/nokia/nperf/blob/master/src/frame_descriptions.rs).)

```rust
pub struct FrameDescriptions< E: Endianity > {
    ...
}

impl< E: Endianity > FrameDescriptions< E > {
    pub fn load( binary: &Arc< BinaryData > ) -> Option< Self > {
        ...
    }

    pub fn find_unwind_info< 'a >( &'a self, ctx_cache: &'a mut ContextCache< E >,
        mappings: &[AddressMapping], address: u64 ) -> Option< UnwindInfo< 'a, E > >
    {
        ...
    }
}

pub struct UnwindInfo< 'a, E: Endianity + 'a > {
    ...
}

impl< 'a, E: Endianity > Drop for UnwindInfo< 'a, E > {
    #[inline]
    fn drop( &mut self ) {
        ... // Cache the context.
    }
}

impl< 'a, E: Endianity > Deref for UnwindInfo< 'a, E > {
    type Target = UnwindTableRow< EndianBuf< 'a, E > >;

    #[inline]
    fn deref( &self ) -> &Self::Target {
        &self.row
    }
}
```

Basically this treat both types of sections in an uniform way, and completely abstracts away the caching of the contexts, which makes it pretty nice to use. You just `load()` the sections, and then call `find_unwind_info`.

**Comment by philipc** [comment]
One possible solution: get rid of the traits and use the same type for both sections, with a boolean parameter to distinguish between the behavior.

**Comment by philipc** [comment]
I've done two changes to address this:
- #399 changed many types to no longer have `Section` type parameters, so you can now treat them the same for both sections.
- #395 changed use of contexts to borrow instead of move, so you can now use them in place. This makes caching contexts simpler.

I think that addresses your main points, but let me know if there is more you think could be done.
